File: flutter_app/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/device.dart';
import '../models/status.dart';

class ApiService {
  static const String _baseUrl = 'http://127.0.0.1:8080';

  Future<Map<String, dynamic>> selectFile(String filePath) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/select-file'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'file_path': filePath}),
    );
    return _handleResponse(response);
  }

  Future<List<DlnaDevice>> discover() async {
    final response = await http.get(Uri.parse('$_baseUrl/api/discover'));
    final data = _handleResponse(response);
    final devices = (data['devices'] as List)
        .map((d) => DlnaDevice.fromJson(d as Map<String, dynamic>))
        .toList();
    return devices;
  }

  Future<void> selectDevice(int deviceIndex) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/select-device'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'device_index': deviceIndex}),
    );
    _handleResponse(response);
  }

  Future<void> cast() async {
    final response = await http.post(Uri.parse('$_baseUrl/api/cast'));
    _handleResponse(response);
  }

  Future<void> play() async {
    final response = await http.post(Uri.parse('$_baseUrl/api/play'));
    _handleResponse(response);
  }

  Future<void> pause() async {
    final response = await http.post(Uri.parse('$_baseUrl/api/pause'));
    _handleResponse(response);
  }

  Future<void> stop() async {
    final response = await http.post(Uri.parse('$_baseUrl/api/stop'));
    _handleResponse(response);
  }

  Future<void> seek(int positionSecs) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/api/seek'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'position_secs': positionSecs}),
    );
    _handleResponse(response);
  }

  Future<PlaybackStatus> getStatus() async {
    final response = await http.get(Uri.parse('$_baseUrl/api/status'));
    final data = _handleResponse(response);
    return PlaybackStatus.fromJson(data);
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    if (response.statusCode >= 400) {
      throw ApiException(body['error'] as String? ?? 'Unknown error');
    }
    return body;
  }
}

class ApiException implements Exception {
  final String message;
  ApiException(this.message);

  @override
  String toString() => message;
}

File: swiftui_app/LocalCast/Services/ApiService.swift
import Foundation

enum ApiError: LocalizedError {
    case serverError(String)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .serverError(let message): return message
        case .networkError(let error): return error.localizedDescription
        }
    }
}

actor ApiService {
    private let baseURL: String
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    init(port: UInt16) {
        self.baseURL = "http://127.0.0.1:\(port)"
    }

    func selectFile(_ filePath: String) async throws -> FileInfo {
        let body = ["file_path": filePath]
        let data = try await post("/api/select-file", body: body)
        return try decoder.decode(FileInfo.self, from: data)
    }

    func discover() async throws -> DiscoverResult {
        let data = try await get("/api/discover")
        return try decoder.decode(DiscoverResult.self, from: data)
    }

    func discoverRefresh() async throws -> DiscoverResult {
        let data = try await post("/api/discover/refresh")
        return try decoder.decode(DiscoverResult.self, from: data)
    }

    func selectDevice(_ deviceIndex: Int) async throws {
        let body = ["device_index": deviceIndex]
        _ = try await post("/api/select-device", body: body)
    }

    func cast() async throws {
        _ = try await post("/api/cast")
    }

    func play() async throws {
        _ = try await post("/api/play")
    }

    func pause() async throws {
        _ = try await post("/api/pause")
    }

    func stop() async throws {
        _ = try await post("/api/stop")
    }

    func seek(_ positionSecs: Int) async throws {
        let body = ["position_secs": positionSecs]
        _ = try await post("/api/seek", body: body)
    }

    func getStatus() async throws -> PlaybackStatus {
        let data = try await get("/api/status")
        return try decoder.decode(PlaybackStatus.self, from: data)
    }

    // MARK: - Private

    private func get(_ path: String) async throws -> Data {
        let url = URL(string: "\(baseURL)\(path)")!
        let (data, response) = try await URLSession.shared.data(from: url)
        try checkResponse(data: data, response: response)
        return data
    }

    private func post(_ path: String, body: Any? = nil) async throws -> Data {
        let url = URL(string: "\(baseURL)\(path)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        try checkResponse(data: data, response: response)
        return data
    }

    private func checkResponse(data: Data, response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if http.statusCode >= 400 {
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let message = json["error"] as? String {
                throw ApiError.serverError(message)
            }
            throw ApiError.serverError("HTTP \(http.statusCode)")
        }
    }
}
